keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,1066 @@
+`metadata.generation` to 1 and to increment `metadata.generation` on update will
+run after all mutating webhooks have finished. 
+
+The pod `metadata.generation` will also continue to be incremented on graceful delete or 
+deferred delete, just as the API server currently does for pods and other 
+objects today.
+
+Pod updates that would not result in `metadata.generation` being incremented include:
+* Changes to metadata (with the exception of `DeletionTimestamp`). This means 
+that if a Pod uses the downward API to make pod metadata available to containers, 
+Pod behavior can change without the generation being incremented. We will consider 
+this working as intended.
+* Changes to status.
+
+##### Client requests to update generation
+
+Any attempts by clients to set or modify the `metadata.generation` field themselves will be
+ignored and overridden by the API Server. This is consistent with existing 
+behavior of the `metadata.generation` field in all other objects. 
+
+#### Kubelet and observedGeneration
+
+When the Kubelet updates the pod status as part of the [pod sync loop](https://github.com/kubernetes/kubernetes/blob/45d0fddaf1f24f7b559eb936308ce2aeb9871850/pkg/kubelet/pod_workers.go#L1214), 
+it will set the `status.observedGeneration` in the pod to reflect the pod `metadata.generation` 
+corresponding to snapshot of the pod currently being synced. That means if the 
+pod spec gets updated concurrently while the kubelet is performing a pod sync loop 
+on a previous update, the `status.observedGeneration` will be behind `metadata.generation`. 
+
+Outside of the pod sync loop, another place where the kubelet sets the pod status is
+when a [pod is rejected (during `HandlePodAdditions`)](https://github.com/kubernetes/kubernetes/blob/8770bd58d04555303a3a15b30c245a58723d0f4a/pkg/kubelet/kubelet.go#L2335). This
+code will also be modified to populate `status.observedGeneration` to express which
+`metadata.generation` was rejected. In this case, the kubelet will not be updating the pod
+through the sync loop (due to the rejection).
+
+The only other place where the pod status is updated is the readiness and probe
+updates, but we will leave `status.observedGeneration` unchanged here as the probe that
+updated the status would be the one synced in the last pod sync loop.
+
+##### Other writers of pod status
+
+There are other writers of pod status besides Kubelet; however this KEP will focus
+solely on Kubelet for the purpose of populating `observedGeneration`. 
+
+Other pod status writers primarily write to pod conditions. Though this KEP proposes to add a `status.conditions[i].observedGeneration` field to the API in addition to the top-level `status.observedGeneration`, 
+we will leave populating `status.conditions[i].observedGeneration` out of scope for this KEP. 
```

Comment:
let's keep it in scope? It shouldn't be a big change

Reply:
I'm not that familiar with the codebase so I can't really comment on how big a change this is. But if you're certain it'll be small, then I'm happy to include it in the scope of this KEP. I've updated the KEP accordingly.  

I just want to make sure we avoid the scenario of risking not getting the `status.observedGeneration` changes into 1.33 because we've coupled it with lower priority changes.

Reply:
we can postpone it to beta

Reply:
perfect, thanks for the suggestion! Moved this to beta under the graduation criteria